File: packages/java/src/main/java/com/slybrowser/AutomationPolicy.java
package com.slybrowser;

import com.microsoft.playwright.Playwright;
import java.util.Collections;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class AutomationPolicy {
  private static final Set<String> SUPPORTED_PLAYWRIGHT_LINES = Collections.singleton("1.61");
  private static final Pattern VERSION = Pattern.compile(
      "^(?<major>[0-9]+)\\.(?<minor>[0-9]+)(?:\\.[0-9]+)?(?:[-+][0-9A-Za-z.-]+)?$");

  private AutomationPolicy() {}

  public static String validatePlaywrightVersion(String version) {
    Matcher match = VERSION.matcher(version);
    if (!match.matches()) {
      throw new ConfigurationException(
          "Invalid Playwright version: " + version,
          "framework_version_invalid");
    }
    String line = match.group("major") + "." + match.group("minor");
    if (!SUPPORTED_PLAYWRIGHT_LINES.contains(line)) {
      throw new ConfigurationException(
          "Unsupported Playwright version " + version + "; supported lines: "
              + String.join(", ", SUPPORTED_PLAYWRIGHT_LINES),
          "framework_version_unsupported");
    }
    return version;
  }

  public static String resolvePlaywrightVersion(String explicitVersion) {
    if (explicitVersion != null) return validatePlaywrightVersion(explicitVersion);
    Package metadata = Playwright.class.getPackage();
    String version = metadata == null ? null : metadata.getImplementationVersion();
    if (version == null || version.trim().isEmpty()) {
      throw new ConfigurationException(
          "Unable to determine the installed com.microsoft.playwright version; pass frameworkVersion explicitly",
          "framework_version_missing");
    }
    return validatePlaywrightVersion(version);
  }

  public static AutomationCapability capability() {
    return capability(AutomationBackend.PROJECT_WEBDRIVER, null);
  }

  public static AutomationCapability capability(AutomationBackend backend, String frameworkVersion) {
    if (backend == AutomationBackend.PROJECT_WEBDRIVER) {
      if (frameworkVersion != null) {
        throw new ConfigurationException(
            "Project WebDriver does not accept a framework version",
            "framework_version_forbidden");
      }
      return new AutomationCapability(backend, "java", null, true, true);
    }
    return new AutomationCapability(
        backend,
        "java",
        resolvePlaywrightVersion(frameworkVersion),
        true,
        true);
  }

  static void requirePlaywrightHumanizeSupport(boolean requested) {
    // Humanize for framework transports is handed to the native browser via
    // --sly-humanize-config. The SDK must not install a language-local substitute.
  }
}

File: packages/dotnet/src/SlyBrowser/AutomationPolicy.cs
using System.Reflection;
using System.Text.RegularExpressions;
using Microsoft.Playwright;

namespace SlyBrowser;

public enum AutomationBackend
{
    ProjectWebDriver,
    Playwright,
}

public sealed record AutomationCapability(
    AutomationBackend Backend,
    string Language,
    string? FrameworkVersion,
    bool NativeHumanize,
    bool PersistentContext);

public static partial class AutomationPolicy
{
    private static readonly HashSet<string> SupportedPlaywrightLines = ["1.61"];

    [GeneratedRegex("^(?<major>[0-9]+)\\.(?<minor>[0-9]+)(?:\\.[0-9]+)?(?:[-+][0-9A-Za-z.-]+)?$")]
    private static partial Regex VersionPattern();

    public static string ValidatePlaywrightVersion(string version)
    {
        Match match = VersionPattern().Match(version);
        if (!match.Success)
            throw new ConfigurationException(
                $"Invalid Playwright version: {version}",
                "framework_version_invalid");
        string line = $"{match.Groups["major"].Value}.{match.Groups["minor"].Value}";
        if (!SupportedPlaywrightLines.Contains(line))
            throw new ConfigurationException(
                $"Unsupported Playwright version {version}; supported lines: {string.Join(", ", SupportedPlaywrightLines)}",
                "framework_version_unsupported");
        return version;
    }

    public static string ResolvePlaywrightVersion(string? explicitVersion = null)
    {
        if (explicitVersion is not null) return ValidatePlaywrightVersion(explicitVersion);
        Assembly assembly = typeof(IPlaywright).Assembly;
        string? version = assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?
            .InformationalVersion
            ?? assembly.GetName().Version?.ToString(3);
        if (string.IsNullOrWhiteSpace(version))
            throw new ConfigurationException(
                "Unable to determine the installed Microsoft.Playwright version",
                "framework_version_missing");
        return ValidatePlaywrightVersion(version);
    }

    public static AutomationCapability Capability(
        AutomationBackend backend = AutomationBackend.ProjectWebDriver,
        string? frameworkVersion = null)
    {
        if (backend is AutomationBackend.ProjectWebDriver)
        {
            if (frameworkVersion is not null)
                throw new ConfigurationException(
                    "Project WebDriver does not accept a framework version",
                    "framework_version_forbidden");
            return new(backend, "dotnet", null, true, true);
        }
        return new(backend, "dotnet", ResolvePlaywrightVersion(frameworkVersion), true, true);
    }

    internal static void RequirePlaywrightHumanizeSupport(bool requested)
    {
        _ = requested;
    }
}
